### PR TITLE
Low level api client

### LINKED ISSUE #462: Abstract SAPI webservice low-level communication from high-level Client flows
Create a higher-level interface to http(s) endpoints in SAPI to abstract away problem(job) submit, status polling, answer download, solver description download, etc.

`Client` can then build on top of that to implement `dwave.cloud.computation.Future` abstraction in a cleaner and simpler way.

In addition, this enables implementing #354, #228, and future SAPI extensions.

### PR DESCRIPTION
**Pull request by randomir**
Closes #462.

### CODE REVIEW

File: dwave/cloud/api/constants.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    """
+
+    PENDING = "PENDING"
+    IN_PROGRESS = "IN_PROGRESS"
+    COMPLETED = "COMPLETED"
+    FAILED = "FAILED"
+    CANCELLED = "CANCELLED"
+
+
+class ProblemEncodingFormat(str, enum.Enum):
```
Comment (arcondello) [inline dwave/cloud/api/constants.py:41]: what is the benefit from subclassing from `str`?

Comment (randomir) [reply dwave/cloud/api/constants.py:41]: JSON serialization works out of the box (treats enum as str), check the [docs](https://docs.python.org/3/library/enum.html#others).

Comment (arcondello) [reply dwave/cloud/api/constants.py:41]: neat!

**Review by arcondello — APPROVED** [review]
LGTM!

File: requirements.txt [hunk 2]
```diff
@@ -1,9 +1,13 @@
 requests[socks]>=2.18
+pydantic>=1.7.3
 homebase>=1.0.0
 python-dateutil>=2.7
 click>=7
 plucky>=0.4.3
 
+# backports
+dataclasses; python_version == "3.7"
```
[the remark this answers is not in the document]
Comment (randomir) [reply requirements.txt:9]: Good catch! Thanks. I got it correct in setup.py, fwiw. :smile:

File: dwave/cloud/api/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+class ResourceLimitsExceededError(RequestError):
+    """Number of resource requests exceed the permitted limit"""
+
+class ResourceBadResponseError(RequestError):
+    """Unexpected resource response"""
+
+class InternalServerError(RequestError):
+    pass
+
```
[the remark this answers is not in the document]
Comment (randomir) [reply dwave/cloud/api/exceptions.py:59]:
```
# This line intentionally left blank.
```
It's to separate HTTP errors from connection (timeout) errors. Not clear, though, I'll admit. :laughing:

**Pull request merged**